Added Linux media integration

pgorgon-hem opened the pull request:
### Summary

Added Linux media integration using MPRIS and `zbus` library. Introduced modular components and  player lifecycle management, metadata handling, timeline synchronization, and playback state updates.

### Key Changes
- Modularized player and session components.
- Introduced methods for accurate position, metadata, and playback state calculations.
- Used player proxies pregenerated by zbus-gen for modular design and consolidated related modules.
- Implemented pure-Rust MPRIS implementation with `zbus` integration to avoid system dependencies.

### Checklist

- [ ] Verify accurate handling of player metadata and timeline updates.
- [ ] Validate compliance with Linux media player standards.
- [ ] Update documentation for the changes made.

pjgorgon committed:
feat(native): add Linux OS watcher and service skeletons

Introduced foundational Linux support for FSCT with a skeleton implementation. Added `run_os_watcher` as a placeholder for MPRIS/D-Bus based media state integration. Implemented `fsct_main` entry point for initializing services, including logging, local driver startup, and graceful shutdown handling.

pjgorgon committed:
feat(native): implement MPRIS-based Linux media integration

Added comprehensive MPRIS integration for Linux, enabling native media state discovery and updates over D-Bus. Updated `run_os_watcher` to register detected players, track state changes, and manage player lifecycle. Documented Linux prerequisites in `README.md` and added necessary dependencies to `Cargo.toml`.

pjgorgon committed:
feat(native): add LoggingDriver example for OS watcher debugging

Introduced a new `LoggingDriver` example for debugging OS watcher integrations. Provides detailed stdout/stderr logging for driver interactions, including player registration, state updates, and lifecycle events. Supports platform-specific OS watchers for Linux, Windows, and macOS.

pjgorgon committed:
feat(native): migrate Linux media integration to pure-Rust MPRIS stack

Replaced `mpris` crate dependency with `zbus` for Linux media integration. Updated player state mapping logic, improved metadata handling, and streamlined `run_os_watcher` implementation. Removed `mpris` and associated system dependencies for cleaner, more portable builds. Updated documentation to reflect changes.

pjgorgon committed:
feat(native): ensure players are unregistered on ownership removal

Added a call to `driver.unregister_player` when player ownership is removed and the registry entry is deleted. Prevents orphaned player entries and ensures proper cleanup during player lifecycle changes.

pjgorgon committed:
feat(linux): improve timeline synchronization and rate handling for MPRIS players

Enhanced timeline synchronization logic by preserving accurate positions and rates during player state changes. Added checks for updated durations and playback statuses to emit precise timeline updates. Improved internal handling of `Position` property and ensured timelines are consistently synchronized with player state.

pjgorgon committed:
feat(native): enhance rate and duration handling for MPRIS players

Improved management of playback rate and duration in MPRIS player state synchronization. Refactored logic for extracting and updating metadata, ensuring accurate timelines during player state or property changes. Added fallback mechanisms for handling rate and duration updates dynamically.

pjgorgon committed:
feat(native): refactor DBus match rule setup in `run_os_watcher`

Moved DBus session initialization and explicit match rule setup outside the service thread in `run_os_watcher`. Ensures early failure detection for connection establishment and match rule configuration. Simplifies the service thread logic by removing redundant rule setup.

pjgorgon committed:
feat(native): abstract MPRIS logic into adapter module

Extracted MPRIS-related operations into a dedicated adapter module for cleaner separation of concerns. Refactored `run_os_watcher` to utilize adapter-provided APIs for match rule installation, initial discovery, and event streaming. Simplified timeline synchronization and metadata handling with centralized helper methods.

pjgorgon committed:
feat(native): simplify MPRIS player integration with `zbus`

Refactored MPRIS integration to utilize the `zbus` library’s autogenerated proxies. Streamlined `run_os_watcher` by delegating player discovery and event handling to a new `PlayerWatcher` abstraction. Removed custom MPRIS event handling code.

pjgorgon committed:
feat(native): refactor player discovery in MPRIS integration

Reworked `iter_player` to streamline MPRIS player discovery using a helper method `get_player`. Simplified handling of NameOwnerChanged signals and removed redundant code for message stream processing. Improved modularity and readability of the player iteration logic.

pjgorgon committed:
feat(native): modularize and refactor MPRIS session watcher

Extracted `SessionWatcher` into a dedicated module to enhance code modularity and readability. Updated `run_os_watcher` to utilize the refactored `SessionWatcher` for player iteration and event handling. Improved separation of concerns in MPRIS player management.

pjgorgon committed:
feat(linux): upgrade zbus integration and simplify MPRIS player handling

Updated `zbus` and related dependencies to their latest versions, enhancing compatibility and feature set. Refactored MPRIS player handling by aligning with updated `zbus` APIs, introducing typed enums for loop and playback statuses, and simplifying proxy generation. Streamlined player lifecycle management and ensured improved code modularity across the Linux media integration.

pjgorgon committed:
feat(native): enhance player registration and lifecycle handling

Introduced `PlayerRegistrationManager` to manage player registration and lifecycle processes, improving modularity and maintainability. Refactored `run_os_watcher` to delegate player event handling to `run_os_player_watcher`, ensuring better separation of concerns. Updated `Player` to use `bus_name` for improved name management and proxy generation consistency.

pjgorgon committed:
feat(linux): restructure MPRIS player proxies and modularize components

Separated the autogenerated `PlayerProxy` code into `media_player2_player` module for improved organization. Refactored `Player` to integrate with the new proxy structure and enhanced interoperability with `SessionWatcher`. Simplified player lifecycle handling by consolidating initialization logic.

pjgorgon committed:
feat(linux): enhance player state handling and playback synchronization

Added `FsctStatus` conversion for `PlaybackStatus` to standardize state updates. Introduced `PlayerHandler` enhancements, including initial state building and playback status change handling. Improved player registration and lifecycle management for better synchronization and state consistency.

pjgorgon committed:
feat(linux): implement detailed player state synchronization and metadata handling

Added methods for recalculating player position, parsing metadata, and handling updated player events (rate, position, metadata). Improved player lifecycle synchronization with enhanced timeline and status updates. Simplified timeline recalibration to ensure accurate playback state representation.

pjgorgon committed:
feat(linux): improve metadata logging in player

Enhanced metadata logging by including detailed information from the parsed metadata map. Provides better traceability and debugging capabilities for metadata change events.

pjgorgon committed:
feat(linux): replace position_changed with seeked signal handling

Replaced `position_changed` signal handling with `seeked` for improved accuracy in playback position updates. Updated the associated task to better manage player state synchronization and handle seek events efficiently.

pjgorgon committed:
feat(linux): add uncached position querying for accurate timeline updates

Introduced `position_uncached` method to bypass proxy caching and retrieve playback position directly via `org.freedesktop.DBus.Properties`. Updated metadata and timeline handling to ensure accurate playback state representation, leveraging uncached position data.

pjgorgon committed:
feat(linux): enhance timeline management and player state updates

Refactored `PlayerHandler` to improve timeline and state management. Introduced `TimelineParts` for better modularity in handling playback position, rate, and duration. Simplified updates for player state, timeline, and metadata to ensure accurate representation and synchronization.

pjgorgon committed:
refactor(linux): consolidate position recalculation methods

Replaced `recalc_position` with `recalculate_position` for improved method naming consistency. Removed the redundant `recalculate_timeline` method, leveraging `recalculate_position` for timeline updates. Simplified player state update flow.

pjgorgon committed:
fix(linux): correct duration type handling in metadata parsing

Updated metadata parsing to prioritize `u64` for `mpris:length` values and added fallback to `i64` handling. Ensures compatibility with unsigned duration values and improves consistency in playback metadata parsing.

pjgorgon committed:
refactor(linux): improve metadata parsing and modularize player components

Refactored metadata parsing by introducing helper methods for text, artists, and duration processing. Improved code readability and modularity. Added missing license headers and consolidated imports in the MPRIS adapter modules. Enhanced maintainability by improving function organization and code structure.

pjgorgon committed:
refactor(linux): replace info logs with debug logs for better granularity

Replaced various `info` log statements with `debug` to reduce verbosity in normal operations and provide more detailed logs for debugging purposes. Added a new debug log for initial player readiness state to enhance traceability.

pjgorgon committed:
refactor: modularize player components and improve logging granularity

Split `PlayerHandler` and `PlayerRegistrationManager` into separate modules for better code organization. Replaced `info` logs with `debug` logs to reduce runtime verbosity while maintaining detailed logs for debugging.

jarekhem commented on ports/native/examples/log_os_watcher.rs:
```diff
@@ -0,0 +1,128 @@
+use fsct_core::player_state::PlayerState;
+use fsct_core::service::ServiceHandle;
+
+#[cfg(target_os = "linux")]
```
To jest dziwne trochę, skoro dla każdego target_os jest to samo.

jarekhem commented on ports/native/src/linux/player/player_handler.rs:
```diff
@@ -0,0 +1,309 @@
+    }
+
+    fn parse_text_array(value: &OwnedValue) -> Option<String> {
+        let multiple_texts: Option<Vec<String>> = value.clone().try_into().ok(); //Vec::<String>::try_from(value).ok();
```
Komentarz

pjgorgon committed:
refactor: remove redundant platform-specific cfg attributes for run_os_watcher

Simplified the code by eliminating redundant `#[cfg(target_os)]` attributes for `run_os_watcher`. Ensures a cleaner and more concise implementation.

pjgorgon committed:
refactor(linux): remove commented code in metadata parsing

Cleaned up redundant commented code in `parse_text_array` method. Improves code readability and maintains a clean codebase.
